type-aliased `seq[T]` get different backend C/C++ pointer type names causing invalid codegen on gcc-14 | g++ | clang++

### Description

```Nim
type
  Uid* = cuint
  Gid* = cuint

proc usrToUid*(usr: string): Uid = 999.Uid

proc usrToUid*(usrs: seq[string]): seq[Uid] =
  for usr in usrs: result.add usrToUid(usr)

proc grpToGid*(grp: string): Gid = 999.Gid

proc grpToGid*(grps: seq[string]): seq[Gid] =
  for grp in grps: result.add grpToGid(grp)

if grpToGid(@["hi"]) != @[999.Gid]: quit 1
```

### Nim Version

nim-devel HEAD from yesterday
```
Nim Compiler Version 2.1.9 [Linux: amd64]
Compiled at 2024-08-03
Copyright (c) 2006-2024 by Andreas Rumpf

active boot switches: -d:release -d:danger -d:nimUseLinenoise
```

### Current Output

The example program fails under `nim cpp --cc=gcc` or `nim cpp --cc=clang` with any backend versions and with `nim c --cc=gcc typeAliasSeq.nim` under gcc-14 (without `--passC=-Wno-error=incompatible-pointer-types` - it works with that).  The error message is roughly the same in all 3 failure cases:

```text
CC: typeAliasSeq.nim
/u/cb/.cache/nim/typeAliasSeq_d/@mtypeAliasSeq.nim.c: In function 'grpToGid__type65lias83eq_u58':
/u/cb/.cache/nim/typeAliasSeq_d/@mtypeAliasSeq.nim.c:151:58: error: passing argument 1 of 'add__type65lias83eq_u30' from incompatible pointer type [-Wincompatible-pointer-types]
  151 |                                 add__type65lias83eq_u30((&result), ((unsigned int) (T5_)));
      |                                                         ~^~~~~~~~
      |                                                          |
      |                                                          tySequence__t9cg37mqWifZWh9b3z1fs0Og *
/u/cb/.cache/nim/typeAliasSeq_d/@mtypeAliasSeq.nim.c:67:93: note: expected 'tySequence__MOLaeFb9cv3vR6ovvXz0FCA *' but argument is of type 'tySequence__t9cg37mqWifZWh9b3z1fs0Og *'
   67 | N_LIB_PRIVATE N_NIMCALL(void, add__type65lias83eq_u30)(tySequence__MOLaeFb9cv3vR6ovvXz0FCA* x_p0, unsigned int y_p1);
      |                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
```

### Expected Output

Should compile and run without errors.  It does so with `nim c --cc=clang`, or with early gcc versions and --cc=gcc, but does NOT with `nim cpp -cc=clang` or later gcc versions.

### Possible Solution

Commenting out the `Uid` type and its two procs makes the problem go away.  This is admittedly more of a "hint" to the solution and relates to my issue title than an actual solution.

### Additional Information

The problem *also* goes away if `cuint` is changed to simply `int` (or `int8` or etc. magic int types).  So, AFAICT, the problem may only apply to `importc` types, but it may also be more broad like generic types, etc.  Worth making a test that tries several things before concluding that it's solved.

--- a/compiler/ccgcalls.nim
+++ b/compiler/ccgcalls.nim
@@ -353,7 +353,7 @@ proc genArg(p: BProc, n: PNode, param: PSym; call: PNode; result: var Rope; need
       addRdLoc(a, result)
   else:
     a = initLocExprSingleUse(p, n)
-    if param.typ.kind in abstractPtrs:
+    if param.typ.kind in {tyVar, tyPtr, tyRef, tySink}:
       let typ = skipTypes(param.typ, abstractPtrs)
       if not sameBackendTypePickyAliases(typ, n.typ.skipTypes(abstractPtrs)):
         a.snippet = "(($1) ($2))" %

--- a/compiler/ccgcalls.nim
+++ b/compiler/ccgcalls.nim
@@ -315,6 +315,12 @@ proc genArgStringToCString(p: BProc, n: PNode; result: var Rope; needsTmp: bool)
   var a = initLocExpr(p, n[0])
   appcg(p.module, result, "#nimToCStringConv($1)", [withTmpIfNeeded(p, a, needsTmp).rdLoc])
 
+# proc filterArgType(t: PType): bool =
+#   case t.kind
+#   of tyEmpty, tyChar, tyBool, tyNil, tyPointer, tyString, tyCstring,
+#      tyInt..tyUInt64, tyTyped, tyUntyped, tyVoid:
+#     result = 
+
 proc genArg(p: BProc, n: PNode, param: PSym; call: PNode; result: var Rope; needsTmp = false) =
   var a: TLoc
   if n.kind == nkStringToCString:
@@ -355,7 +361,7 @@ proc genArg(p: BProc, n: PNode, param: PSym; call: PNode; result: var Rope; need
     a = initLocExprSingleUse(p, n)
     if param.typ.kind in abstractPtrs:
       let typ = skipTypes(param.typ, abstractPtrs)
-      if typ.sym != nil and sfImportc in typ.sym.flags:
+      if not sameBackendTypePickyAliases(typ, n.typ.skipTypes(abstractPtrs)):
         a.snippet = "(($1) ($2))" %
           [getTypeDesc(p.module, param.typ), rdCharLoc(a)]
     addRdLoc(withTmpIfNeeded(p, a, needsTmp), result)

--- a/compiler/ccgcalls.nim
+++ b/compiler/ccgcalls.nim
@@ -315,12 +315,6 @@ proc genArgStringToCString(p: BProc, n: PNode; result: var Rope; needsTmp: bool)
   var a = initLocExpr(p, n[0])
   appcg(p.module, result, "#nimToCStringConv($1)", [withTmpIfNeeded(p, a, needsTmp).rdLoc])
 
-# proc filterArgType(t: PType): bool =
-#   case t.kind
-#   of tyEmpty, tyChar, tyBool, tyNil, tyPointer, tyString, tyCstring,
-#      tyInt..tyUInt64, tyTyped, tyUntyped, tyVoid:
-#     result = 
-
 proc genArg(p: BProc, n: PNode, param: PSym; call: PNode; result: var Rope; needsTmp = false) =
   var a: TLoc
   if n.kind == nkStringToCString:

--- a/compiler/types.nim
+++ b/compiler/types.nim
@@ -984,7 +984,7 @@ type
     AllowCommonBase
     PickyCAliases  # be picky about the distinction between 'cint' and 'int32'
     IgnoreFlags    # used for borrowed functions and methods; ignores the tfVarIsPtr flag
-    PickyBackendAliases # be picky different aliases
+    PickyBackendAliases # be picky about different aliases
 
   TTypeCmpFlags* = set[TTypeCmpFlag]
 

--- a/compiler/types.nim
+++ b/compiler/types.nim
@@ -984,6 +984,7 @@ type
     AllowCommonBase
     PickyCAliases  # be picky about the distinction between 'cint' and 'int32'
     IgnoreFlags    # used for borrowed functions and methods; ignores the tfVarIsPtr flag
+    PickyBackendAliases # be picky different aliases
 
   TTypeCmpFlags* = set[TTypeCmpFlag]
 
@@ -1260,6 +1261,11 @@ proc sameTypeAux(x, y: PType, c: var TSameTypeClosure): bool =
       let symFlagsB = if b.sym != nil: b.sym.flags else: {}
       if (symFlagsA+symFlagsB) * {sfImportc, sfExportc} != {}:
         result = symFlagsA == symFlagsB
+    elif result and PickyBackendAliases in c.flags:
+      let symFlagsA = if a.sym != nil: a.sym.flags else: {}
+      let symFlagsB = if b.sym != nil: b.sym.flags else: {}
+      if (symFlagsA+symFlagsB) * {sfImportc, sfExportc} != {}:
+        result = a.id == b.id
 
   of tyStatic, tyFromExpr:
     result = exprStructuralEquivalent(a.n, b.n) and sameFlags(a, b)
@@ -1343,6 +1349,12 @@ proc sameTypeAux(x, y: PType, c: var TSameTypeClosure): bool =
 proc sameBackendType*(x, y: PType): bool =
   var c = initSameTypeClosure()
   c.flags.incl IgnoreTupleFields
+  c.cmp = dcEqIgnoreDistinct
+  result = sameTypeAux(x, y, c)
+
+proc sameBackendTypePickyAliases*(x, y: PType): bool =
+  var c = initSameTypeClosure()
+  c.flags.incl {IgnoreTupleFields, PickyCAliases, PickyBackendAliases}
   c.cmp = dcEqIgnoreDistinct
   result = sameTypeAux(x, y, c)
 

